File: Sorting/mergeSort.swift
#!/usr/bin/env xcrun swift
func mergeSort(_ array: [Int]) -> [Int] {
    guard array.count > 1 else { return array }
    let middleIndex = array.count / 2
    let leftArray = mergeSort(Array(array[0..<middleIndex]))
    let rightArray = mergeSort(Array(array[middleIndex..<array.count]))
    return merge(leftArray: leftArray, rightArray: rightArray)
}

private func merge(leftArray: [Int], rightArray: [Int]) -> [Int] {
    var i = 0
    var j = 0
    var ordered = [Int]()
    while i < leftArray.count && j < rightArray.count {
        if leftArray[i] < rightArray[j] {
        ordered.append(leftArray[i])
        i += 1
    } else {
        ordered.append(rightArray[j])
        j += 1
    } 
    }
    while i < leftArray.count {
        ordered.append(leftArray[i])
        i += 1
    }
    while j < rightArray.count {
        ordered.append(rightArray[j])
        j += 1
    }
    return ordered
}

let array = [3, 5, 24, 6, 2, 4, 8, 7]
print(mergeSort(array))

File: Sorting/merge_sort.py
#!/usr/bin/env python


def merge_sort(list):
    mid = len(list) // 2
    if len(list) > 1:
        left = []
        right = []
        for index in range(0, mid):
            left.append(list[index])
        for index in range(mid, len(list)):
            right.append(list[index])
        merge_sort(left)
        merge_sort(right)
        merge(left, right, list)
        return list


def merge(left, right, list):
    i = 0
    j = 0
    k = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            list[k] = left[i]
            i += 1
        else:
            list[k] = right[j]
            j += 1
        k += 1
    while i < len(left):
        list[k] = left[i]
        i += 1
        k += 1
    while j < len(right):
        list[k] = right[j]
        j += 1
        k += 1


list = [3, 5, 24, 6, 2, 4, 8, 7]
print merge_sort(list)
